x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java

```diff
@@ -0,0 +1,100 @@
+/*
+ *
+ *  Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ *  or more contributor license agreements. Licensed under the Elastic License;
+ *  you may not use this file except in compliance with the Elastic License.
+ *
+ */
+
+package org.elasticsearch.xpack.core.security.authz.privilege;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.xpack.core.security.action.CreateApiKeyRequest;
+import org.elasticsearch.xpack.core.security.action.GetApiKeyRequest;
+import org.elasticsearch.xpack.core.security.action.InvalidateApiKeyRequest;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authz.permission.ClusterPermission;
+import org.elasticsearch.xpack.core.security.support.Automatons;
+
+/**
+ * Named cluster privilege for managing API keys owned by the current authenticated user.
+ */
+public class ManageOwnApiKeyClusterPrivilege implements NamedClusterPrivilege {
+    public static final ManageOwnApiKeyClusterPrivilege INSTANCE = new ManageOwnApiKeyClusterPrivilege();
+    private static final String PRIVILEGE_NAME = "manage_own_api_key";
+
+    private ManageOwnApiKeyClusterPrivilege() {
+    }
+
+    @Override
+    public String name() {
+        return PRIVILEGE_NAME;
+    }
+
+    @Override
+    public ClusterPermission.Builder buildPermission(ClusterPermission.Builder builder) {
+        return builder.add(this, ManageOwnClusterPermissionCheck.INSTANCE);
+    }
+
+    private static final class ManageOwnClusterPermissionCheck extends ClusterPermission.ActionBasedPermissionCheck {
+        public static final ManageOwnClusterPermissionCheck INSTANCE = new ManageOwnClusterPermissionCheck();
+
+        private ManageOwnClusterPermissionCheck() {
+            super(Automatons.patterns("cluster:admin/xpack/security/api_key/*"));
+        }
+
+        @Override
+        protected boolean doCheck(String action, TransportRequest request, Authentication authentication) {
+            if (request instanceof CreateApiKeyRequest) {
+                return true;
+            } else if (request instanceof GetApiKeyRequest) {
+                final GetApiKeyRequest getApiKeyRequest = (GetApiKeyRequest) request;
+                return checkIfUserIsOwnerOfApiKeys(authentication, getApiKeyRequest.getApiKeyId(), getApiKeyRequest.getUserName(),
+                    getApiKeyRequest.getRealmName());
+            } else if (request instanceof InvalidateApiKeyRequest) {
+                final InvalidateApiKeyRequest invalidateApiKeyRequest = (InvalidateApiKeyRequest) request;
+                return checkIfUserIsOwnerOfApiKeys(authentication, invalidateApiKeyRequest.getId(),
+                    invalidateApiKeyRequest.getUserName(),
+                    invalidateApiKeyRequest.getRealmName());
```

Comment:
Shouldn't this be looking at `ownedByAuthenticatedUser` ?

Reply:
I'm sorry, I don't understand how that's supposed to work.
If I have `manage_own_api_key` then I should be able to call `GET /_security/api_key?owner=true` correct? That's _why_ we added the owner flag.

But this implementation requires that I populate realm and username.

Reply:
@tvernum But the transport actions sets the username and realm when the `owner=true`:
```
            if (authentication.getAuthenticatedBy().getType().equals(ApiKeyService.API_KEY_REALM_TYPE)) {
                realm = (String) authentication.getMetadata().get(ApiKeyService.API_KEY_CREATOR_REALM);
            } else {
                realm = authentication.getAuthenticatedBy().getName();
            }
```
So it is not required to check the owner flag in authorization, and just check that the username and realm in the request match the ones of the currently authenticated user. The `APIKeyService` builds the search filter based on the username and realm, and not on the owner flag, so those are the parameters we should authorize. Ideally, the owner flag should not even be present in the transport request, it was my preference to keep this in the "REST" land.

At this stage I'm happy with whatever option, but in my mind it is simpler to not use the `owner` flag but in the REST layer to build a transport request "correctly" populated.

Reply:
@albertzaharovits As the rest of the code stands now, the privilege needs to check `owner` because it's handled in the transport layer, not the rest layer. The username+realm fields are not populated at the time the privilege is checked.  
We can talk about moving it to the rest layer instead, but as it stands, this PR is wrong if it doesn't deal with it.

Originally, it was my preference to have `owner` handled in the rest layer too, I almost pushed for that when we reviewed that PR, but it changes a few things.
To do that we need to move `owner` in the invalidate (`DELETE`) API to be a URL parameter and not be in the request body. If it's in the body, then it's in the request object, and if it's in the request object, then it really needs to be handled by the transport action.
That then shifts some of the validation responsibilities because the Rest action needs to ensure that you don't set `owner` and a username/realm.
It's all possible, but it means we're back-tracking on previous PRs.

_This_ PR needs to work with the implementation that's there, and that means that it has to validate on the `owner` flag, or the owner flag becomes totally useless.

Reply:
> Because of backward compatibility ...

This privilege doesn't exist in any older versions. I fail to see how BWC comes into this. I'm happy to support validation on username+realm, but it's not a BWC matter.